gu-client/examples/gu-client.rs

```diff
@@ -0,0 +1,1017 @@
+            MainSteps::UploadFile {
+                file_name,
+                total_upload_progress: Progress::Step(s, _),
+            } => {
+                self.upload_bar
+                    .message(&format!("upload [{}] ", file_name.display()));
+                self.upload_bar.set(s);
+            }
+            MainSteps::UploadFile {
+                file_name,
+                total_upload_progress: Progress::Done,
+            } => {
+                self.upload_bar.finish_println("upload done");
+                self.main_bar.add(1);
+            }
+            MainSteps::Frame(frame) => {
+                self.main_bar.message(&format!("[frame {}]:  ", frame));
+                self.main_bar.add(1);
+            }
+            MainSteps::Done => {
+                self.main_bar.finish_println("rendering done");
+            }
+            e => {
+                //eprintln!("!!event={:?}", e);
+            }
+        }
+    }
+}
+
+fn parse_resolution(r: &str) -> Result<(u32, u32), failure::Error> {
+    let mut it = r.split("x");
+
+    Ok(match (it.next(), it.next(), it.next()) {
+        (Some(w), Some(h), None) => (w.parse()?, h.parse()?),
+        _ => Err(Error::Other(format!("invalid format {}", r)))?,
+    })
+}
+
+#[derive(Debug, Clone)]
+enum FrameRange {
+    Single(u32),
+    Range { from: u32, to: u32, step_by: u32 },
+}
+
+impl Iterator for FrameRange {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (item, next) = match self {
+            FrameRange::Single(v) => (
+                Some(*v),
+                FrameRange::Range {
+                    from: *v + 1,
+                    to: *v,
+                    step_by: 1,
+                },
+            ),
+            FrameRange::Range { from, to, step_by } => {
+                if from <= to {
+                    (
+                        Some(*from),
+                        FrameRange::Range {
+                            from: (*from) + (*step_by),
+                            to: (*to),
+                            step_by: *step_by,
+                        },
+                    )
+                } else {
+                    (
+                        None,
+                        FrameRange::Range {
+                            from: *from,
+                            to: *to,
+                            step_by: *step_by,
+                        },
+                    )
+                }
+            }
+        };
+        *self = next;
+        item
+    }
+}
+
+fn parse_frame_range(r: &str) -> Result<Vec<FrameRange>, failure::Error> {
+    let mut it = r.chars().peekable();
+    let mut r = Vec::new();
+
+    fn peek_range<T: Iterator<Item = char>>(
+        it: &mut Peekable<T>,
+    ) -> Result<Option<FrameRange>, Error> {
+        let left = match shift_number(it) {
+            Some(n) => n,
+            None => return Ok(None),
+        };
+        if it.peek() != Some(&'-') {
```

Comment:
Instead of peeking we can call next() right away.

```suggestion
        if it.next() != Some(&'-') {
```

Reply:
it breaks test:

```
assert_eq!(
            parse_frame_range("1-5,3,4,7").unwrap(),
            vec![
                FrameRange::Range {
                    from: 1,
                    to: 5,
                    step_by: 1
                },
                FrameRange::Single(3),
                FrameRange::Single(4),
                FrameRange::Single(7)
            ]
)
```